lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp

```diff
@@ -2584,37 +2586,104 @@ lldb::addr_t SwiftLanguageRuntime::GetAsyncContext(RegisterContext *regctx) {
   return LLDB_INVALID_ADDRESS;
 }
 
-/// Creates an expression accessing *(fp - 8) or **(fp - 8) if
-/// `with_double_deref` is true. This is only valid for x86_64 or aarch64.
-llvm::ArrayRef<uint8_t>
-GetAsyncRegFromFramePointerDWARFExpr(llvm::Triple::ArchType triple,
-                                     bool with_double_deref) {
-  assert(triple == llvm::Triple::x86_64 || triple == llvm::Triple::aarch64);
-
-  // These expressions must have static storage, due to how UnwindPlan::Row
-  // works.
-  static const uint8_t g_cfa_dwarf_expression_x86_64[] = {
-      llvm::dwarf::DW_OP_breg6, // DW_OP_breg6, register 6 == rbp
-      0x78,                     //    sleb128 -8 (ptrsize)
-      llvm::dwarf::DW_OP_deref,
-      llvm::dwarf::DW_OP_deref,
-  };
-  static const uint8_t g_cfa_dwarf_expression_arm64[] = {
-      llvm::dwarf::DW_OP_breg29, // DW_OP_breg29, register 29 == fp
-      0x78,                      //    sleb128 -8 (ptrsize)
-      llvm::dwarf::DW_OP_deref,
-      llvm::dwarf::DW_OP_deref,
-  };
+/// Functional wrapper to read a register as an address.
+std::optional<addr_t> ReadRegisterAsAddress(RegisterContext &regctx,
+                                            unsigned regnum) {
+  auto reg = regctx.ReadRegisterAsUnsigned(regnum, LLDB_INVALID_ADDRESS);
+  if (reg != LLDB_INVALID_ADDRESS)
+    return reg;
+  return {};
+}
+
+/// Functional wrapper to read a pointer from process memory at `addr +
+/// offset`.
+std::optional<addr_t>
+ReadPtrFromAddr(Process &process, std::optional<addr_t> addr, int offset = 0) {
+  if (!addr)
+    return {};
+  Status error;
+  addr_t ptr = process.ReadPointerFromMemory(*addr + offset, error);
+  if (ptr != LLDB_INVALID_ADDRESS)
+    return ptr;
+  return {};
+}
+
+/// Converts the abstract location of UnwindPlan::Row::FAValue into a concrete
+/// address. This is a simplified version of the methods in
+/// RegisterContextUnwind, since plumbing access to those here would be
+/// challenging.
+std::optional<addr_t> GetCFAAddr(Process &process, RegisterContext &regctx,
```

Comment:
This method will need to take the UnwindPlan's `m_register_kind` as an argument.  

When you call `RegisterContext::ReadRegisterAsUnsigned`, it takes a register number in the `eRegisterKindLLDB` numbering system.  You can use `RegisterContext::ConvertRegisterKindToRegisterNumber` to convert from the UnwindPlan's register numbering scheme to `eRegisterKindLLDB` before passing it to `ReadRegisterAsUnsigned`.  e.g. see `UnwindPlan::GetRegisterInfo`.

The UnwindPlan's use the native register numbering of their input sources, e.g. eh_frame numbering or DWARF debug_frame, or the ISA's natural numbering scheme.  They're usually the same as lldb's numbering system for a given target, but it's not guaranteed and can result in subtle bugs.

(Yes, it's a poor design that should really use an object which tracks number & numbering domain)

Reply:
This is exactly the kind of feedback I needed! I'll work on it next

Reply:
Instead of passing the `m_register_kind` as an argument, you could have this method require that the register number be in eRegisterKindLLDB already, and the caller must do the conversion before calling.  that's probably simpler.

Reply:
I tried both approaches, but I think the "argument" idea is a bit better (although more work) because it makes it harder to misuse, since people will need to manually pass in a register kind (and therefore think about register domains)